components/workspace-rollout-job/pkg/rollout/rollout.go

```diff
@@ -0,0 +1,184 @@
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	logrus "github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/workspace-rollout-job/pkg/analysis"
+	"github.com/gitpod-io/gitpod/workspace-rollout-job/pkg/wsbridge"
+)
+
+type RollOutJob struct {
+	oldCluster           string
+	newCluster           string
+	currentScore         int32
+	okayScoreUntilNoData int32
+	analyzer             analysis.Analyzer
+	RolloutAction        wsbridge.RolloutAction
+	rolloutStep          int32
+	analysisWaitDuration time.Duration
+
+	ticker *time.Ticker
+	revert chan bool
+	done   chan bool
+}
+
+func New(oldCluster, newCluster string, rolloutWaitDuration, analysisWaitDuration time.Duration, step, okayScoreUntilNoData int32, analyzer analysis.Analyzer, rolloutAction wsbridge.RolloutAction) *RollOutJob {
+	return &RollOutJob{
+		oldCluster:           oldCluster,
+		newCluster:           newCluster,
+		okayScoreUntilNoData: okayScoreUntilNoData,
+		currentScore:         0,
+		rolloutStep:          step,
+		analyzer:             analyzer,
+		RolloutAction:        rolloutAction,
+		done:                 make(chan bool, 1),
+		revert:               make(chan bool, 1),
+		analysisWaitDuration: analysisWaitDuration,
+		// move forward every waitDuration
+		ticker: time.NewTicker(rolloutWaitDuration),
+	}
+}
+
+// Start runs the job synchronously
+func (r *RollOutJob) Start(ctx context.Context) error {
+	log := logrus.WithField("component", "rollout-job")
+	log.Infof("Rollout job started with the following configuration: %+v", r)
+
+	// Handle interrupt signal
+	c := make(chan os.Signal, 1)
+	signal.Notify(c,
+		os.Interrupt,
+		syscall.SIGTERM,
+	)
+	go func() {
+		<-c
+		// sig is a ^C, handle it
+		logrus.Info("Received interrupt signal, Reverting")
+		r.revert <- true
+	}()
+
+	// keep checking the analyzer asynchronously to see if there is a
+	// problem with the new cluster
+	go func() {
+		for {
+			// Run only if the revert channel is empty
+			if len(r.revert) == 0 {
+				// check every analysisWaitDuration
+				time.Sleep(r.analysisWaitDuration)
+
+				moveForward, err := r.analyzer.MoveForward(context.Background(), r.newCluster)
+				if err != nil {
+					log.Error("Analysis returned error: ", err)
+					log.Info("Reverting the rollout")
+					// Revert the rollout in case of analysis failure
+					r.revert <- true
+					return
+				}
+
+				// Analyzer says no, stop the rollout
+				if moveForward == -1 {
+					log.Info("Analyzer says no, stopping the rollout")
+					r.revert <- true
+					return
+				} else if moveForward == 0 {
+					log.Info("Analyzer says no data, waiting for more data")
+					// Rollout okay until currentScore < okayScoreUntilNoData
+					if r.currentScore >= r.okayScoreUntilNoData {
+						log.Info("Current score is more than okayScoreUntilNoData, reverting the rollout")
+						r.revert <- true
+						return
+					}
+				} else {
+					// Roll forward otherwise
+					log.Info("Analyzer says yes, rolling forward")
+				}
```

Comment:
We could make code a little bit more readable if we decrease the indentation level 🙂:

```go
// Analyzer says no, stop the rollout
if moveForward == -1 {
	log.Info("Analyzer says no, stopping the rollout")
	r.revert <- true
	return
} else if moveForward == 0 && if r.currentScore >= r.okayScoreUntilNoData {
	log.Info("Current score is more than okayScoreUntilNoData, reverting the rollout")
	r.revert <- true
	return
}
// Roll forward otherwise
log.Info("Analyzer says yes, rolling forward")

```